include/cuco/detail/static_map.inl

```diff
@@ -192,33 +224,59 @@ __device__ bool static_map<Key, Value, Scope, Allocator>::device_mutable_view::i
       uint32_t src_lane = __ffs(window_contains_empty) - 1;
 
       if (g.thread_rank() == src_lane) {
-        using cuda::std::memory_order_relaxed;
         auto expected_key   = this->get_empty_key_sentinel();
         auto expected_value = this->get_empty_value_sentinel();
-        auto& slot_key      = current_slot->first;
-        auto& slot_value    = current_slot->second;
-
-        bool key_success =
-          slot_key.compare_exchange_strong(expected_key, insert_pair.first, memory_order_relaxed);
-        bool value_success = slot_value.compare_exchange_strong(
-          expected_value, insert_pair.second, memory_order_relaxed);
-
-        if (key_success) {
-          while (not value_success) {
-            value_success =
-              slot_value.compare_exchange_strong(expected_value = this->get_empty_value_sentinel(),
-                                                 insert_pair.second,
-                                                 memory_order_relaxed);
+
+        // Packed CAS for 4B/4B key/value pairs
+        if constexpr (sizeof(Key) == 4 and sizeof(Value) == 4) {
+          pair2uint64 converter;
+          auto slot = reinterpret_cast<cuda::atomic<uint64_t>*>(current_slot);
+
+          converter.pair =
+            cuco::make_pair<Key, Value>(std::move(expected_key), std::move(expected_value));
+          auto empty_sentinel = converter.uint64;
+          converter.pair      = insert_pair;
+          auto tmp_pair       = converter.uint64;
+
+          bool success =
+            slot->compare_exchange_strong(empty_sentinel, tmp_pair, memory_order_relaxed);
+          if (success) {
+            status = insert_result::SUCCESS;
+          }
+          // duplicate present during insert
+          else if (key_equal(insert_pair.first, current_slot->first.load(memory_order_relaxed))) {
+            status = insert_result::DUPLICATE;
           }
-          status = insert_result::SUCCESS;
-        } else if (value_success) {
-          slot_value.store(this->get_empty_value_sentinel(), memory_order_relaxed);
         }
+        // Back-to-back CAS for 8B/8B key/value pairs
+        else {
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 700
+#error "8B/8B key/value pair insertions are only supported for sm_70 and up."
+#endif
```

Comment:
This isn't going to work. `#error` doesn't respect `if constexpr` branches. e.g., https://godbolt.org/z/q11b11zq7

Reply:
I think it'd be better to do something like: 

`static_assert( __CUDA_ARCH__ > 700 or (sizeof(value_type) <= 8));`

This can go in the class declaration instead of inside the function here.

Reply:
> This isn't going to work. `#error` doesn't respect `if constexpr` branches. e.g., https://godbolt.org/z/q11b11zq7

I thought error directive is always used within a `#ifdef` scope. This will work: https://godbolt.org/z/adjT3f13n

Reply:
Yes, but imagine the case where `defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 700` is true. The `#error` will always fire, even though it takes the `if` branch instead of the `else`.

Reply:
OK, I see.